tests/unit/system/relationships/relationship-payload-manager-test.js

```diff
@@ -674,3 +678,72 @@ test('get can retrieve payloads with self-links in reflexive relationships', fun
   let entry = this.relationshipPayloadsManager.get('user', 1, 'friends');
   assert.deepEqual(entry, friendsPayload, 'self-link in reflexive relationship');
 });
+
+test('handles relationships where one side is polymorphic', function(assert) {
```

Comment:
as discussed out of band, need more tests

- push subtype then push basetype
- push basetype then push subtype
- both sides poly

Reply:
Any way I can keep this moving?  Is it possible for me to merge tests into this feature branch? Never tried that before

Reply:
@danielspaniel you can open pull requests against runspired/data to branch `fix-polymporphic-payloads` and when they're merged it'll update this pr.  I've done it before and it works surprisingly well.

@danielspaniel if you have time to contribute tests for this that would be super helpful.  Thanks ^_^

Reply:
I have all tests ready but @runspired  added this to RelationshipPayloads constructor: 
```
assert(`Both side of a relationship cannot be polymorphic`, !this._lhsIsPolymorphic || !this._rhsIsPolymorphic);
```

That is showstopper, since previously in ember data you could do that .. so I think that assertion is perhaps guarding agains something I don't know what.  

But I removed that assertion and the test failed for both sides polymorphic, so maybe that had something to do with it .. not sure

Reply:
This feature ( lazy relationship loading ) is super complex and was implemented with very limited tests in the first place.  
 I am thinking it might be better to dump the feature,  rather than keep patching it.

  Was this your feature @hjdivad ?  Is there a great value to it.  If there is I am not seeing it. 
Would you be up for dumping it/ removing it?

Reply:
So, if a feature is not done properly and introduces very serious bugs,  are you telling me you are reluctant to back out the feature because you would lose the nice performance gains? 
 That makes no sense. 
First off, whoever wrote the feature should take responsibility and either fix it or remove it. 
I am willing to help out as much as I can, but the way it was done really looks like spaghetti to me, so I can't do much to save it. 
Secondly, if I wrote this feature, I would be embarrassed to keep it, if it broke ember-data. Does not really matter what kind of fun it offers if it is broken.  

How long should this broken feature sit in ember-data and just kill off polymorphic relationship users? 
Is it right to just wait around for some other work someone else is doing?  Who knows how long that other thing Igor is working on will be done.  Seems irrelevant to me, especially since he is not involved in this discussion much. It has only been you, me , Chris and Stephan ( saying he liked the original bug posting  )

Reply:
@danielspaniel Sorry for being a bit quiet. Last week when @hjdivad and I discussed this PR OOB we discussed that assert.  Polymorphic to Polymorphic in ED has always been very buggy, and I wasn't sure it really worked at all and I thought it'd be too difficult to achieve here; however, in that conversation David and I also arrived at a solution for supporting it. So long as we always resolve to the baseModelName on both sides, the approach this PR outlines will work for both sides being polymorphic as well.  I will update the PR with that approach.

Reply:
@danielspaniel can you PR the tests you've added to my branch?

Reply:
sure thing .. gimme a few moments

Reply:
I submitted PR to your branch of ED just now